### PR TITLE
Shared library support added

### PR DESCRIPTION
**Pull request by smashra**
Once the following errors are resolved shared library will be built fine.

Locally the build can be tested with shared library by:

./configure  --disable-kex-sidh-msr; make clean; make -j8; make test 

Static library can be tested by:

./configure  LDFLAGS="-static" --disable-kex-sidh-msr; make clean; make -j8; make test 

src/kem/sike/.libs/libkemsike.a(libkemsike_la-P503.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kem/sike/P503/P503.c:22: multiple definition of `p503'
src/kex_sidh_msr/.libs/libsidhmsr.a(libsidhmsr_la-P503.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kex_sidh_msr/P503/P503.c:22: first defined here
src/kem/sike/.libs/libkemsike.a(libkemsike_la-P503.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kem/sike/P503/P503.c:24: multiple definition of `p503p1'
src/kex_sidh_msr/.libs/libsidhmsr.a(libsidhmsr_la-P503.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kex_sidh_msr/P503/P503.c:24: first defined here
src/kem/sike/.libs/libkemsike.a(libkemsike_la-P503.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kem/sike/P503/P503.c:26: multiple definition of `p503x2'
src/kex_sidh_msr/.libs/libsidhmsr.a(libsidhmsr_la-P503.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kex_sidh_msr/P503/P503.c:26: first defined here
src/kem/sike/.libs/libkemsike.a(libkemsike_la-P751.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kem/sike/P751/P751.c:22: multiple definition of `p751'
src/kex_sidh_msr/.libs/libsidhmsr.a(libsidhmsr_la-P751.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kex_sidh_msr/P751/P751.c:22: first defined here
src/kem/sike/.libs/libkemsike.a(libkemsike_la-P751.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kem/sike/P751/P751.c:24: multiple definition of `p751p1'
src/kex_sidh_msr/.libs/libsidhmsr.a(libsidhmsr_la-P751.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kex_sidh_msr/P751/P751.c:24: first defined here
src/kem/sike/.libs/libkemsike.a(libkemsike_la-P751.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kem/sike/P751/P751.c:26: multiple definition of `p751x2'
src/kex_sidh_msr/.libs/libsidhmsr.a(libsidhmsr_la-P751.o):/home/s24mishra/workspace/alt_liboqs/liboqs/src/kex_sidh_msr/P751/P751.c:26: first defined here
collect2: error: ld returned 1 exit status
Makefile:628: recipe for target 'liboqs.la' failed

### CODE REVIEW

**Comment by dstebila** [comment]
@smashra once #368 lands can you revisit this?

**Comment by dstebila** [comment]
#368 has landed so this should be able to resolved, @smashra

File: config/flags.m4 [hunk 1]
```diff
@@ -22,7 +22,7 @@ AC_DEFUN([SET_AM_CFLAGS],
   AM_CFLAGS=${AM_CFLAGS}" -fPIC "
   
   # Hidding all unneeded symbols.
-  AM_CFLAGS=${AM_CFLAGS}" -fvisibility=hidden "
+  #AM_CFLAGS=${AM_CFLAGS}" -fvisibility=hidden "
```
Comment (dstebila) [inline config/flags.m4]: Why has this been removed?

Comment (smashra) [reply config/flags.m4]: Some symbols were not visible when building the shared library.
On make test I was getting following errors:
Undefined symbols for architecture x86_64:
  "_OQS_KEM_frodokem_640_aes_decaps", referenced from:
      _example_stack in example_kem.o
  "_OQS_KEM_frodokem_640_aes_encaps", referenced from:
      _example_stack in example_kem.o
  "_OQS_KEM_frodokem_640_aes_keypair", referenced from:
      _example_stack in example_kem.o
  "_OQS_MEM_cleanse", referenced from:
      _example_stack in example_kem.o
      _cleanup_stack in example_kem.o
  "_OQS_MEM_insecure_free", referenced from:
      _example_heap in example_kem.o
      _cleanup_heap in example_kem.o
  "_OQS_MEM_secure_free", referenced from:
      _example_heap in example_kem.o
      _cleanup_heap in example_kem.o

Comment (dstebila) [reply config/flags.m4]: Can you investigate to see the cause of the individual errors?  If memory serves we turned `-fvisibility=hidden` on for a reason, to make sure we were exporting exactly the symbols we wanted to export.  So if that's not the case, I'd rather fix it than just turn the feature off.

**Comment by smashra** [comment]
Looked at it.
It's not that straightforward as far as I can tell.
The code will have to follow the integration pattern of bike.

**Comment by smashra** [comment]
Looks Like I have to put
OQS_API defined as (__attribute__((visibility("default")))
in front of lots of functions.

**Comment by smashra** [comment]
List of functions:

src/common/common.h:void OQS_API OQS_print_hex_string(const char *label, const uint8_t *str, size_t len);
src/common/common.h:void OQS_API OQS_print_part_hex_string(const char *label, const uint8_t *str, size_t len, size_t sub_len);
src/common/common.h:void OQS_API OQS_MEM_cleanse(void *ptr, size_t len);
src/common/common.h:void OQS_API OQS_MEM_secure_free(void *ptr, size_t len);
src/common/common.h:void OQS_API OQS_MEM_insecure_free(void *ptr);
src/common/rand.h:OQS_STATUS OQS_API OQS_randombytes_switch_algorithm(const char *algorithm);
src/common/rand.h:void OQS_API OQS_randombytes(uint8_t *random_array, size_t bytes_to_read);
src/common/rand.h:void OQS_API OQS_randombytes_nist_kat_init(const uint8_t *entropy_input, const uint8_t *personalization_string, int security_strength);
src/common/rand.h:OQS_RAND OQS_API *OQS_RAND_new(enum OQS_RAND_alg_name alg_name);
src/common/rand.h:uint8_t OQS_API OQS_RAND_8(OQS_RAND *r);
src/common/rand.h:uint32_t OQS_API OQS_RAND_32(OQS_RAND *r);
src/common/rand.h:uint64_t OQS_API OQS_RAND_64(OQS_RAND *r);
src/common/rand.h:void OQS_API OQS_RAND_n(OQS_RAND *r, uint8_t *out, size_t n);
src/common/rand.h:void OQS_API OQS_RAND_free(OQS_RAND *r);
src/common/rand.h:void OQS_API OQS_RAND_test_record_occurrence(const unsigned char b, unsigned long occurrences[256]);
src/common/rand.h:double OQS_API OQS_RAND_test_statistical_distance_from_uniform(const unsigned long occurrences[256]);
src/common/rand.h:double OQS_API OQS_RAND_zscore_deviation_from_uniform(const unsigned long occurrences[256]);
src/common/rand.h:void OQS_API OQS_RAND_report_statistics(const unsigned long occurrences[256], const char *indent);
src/common/rand.h:OQS_STATUS OQS_API OQS_RAND_get_system_entropy(uint8_t *buf, size_t n);
src/common/rand_nist.c:void OQS_API OQS_randombytes_nist_kat_init(unsigned char *entropy_input, unsigned char *personalization_string, int security_strength) {
src/crypto/aes/aes_c.c:void OQS_API oqs_aes128_load_schedule_c(const uint8_t *key, void **_schedule) {
src/crypto/aes/aes_c.c:void OQS_API oqs_aes128_free_schedule_c(void *schedule) {
src/crypto/aes/aes_c.c:void OQS_API oqs_aes256_load_schedule_c(const uint8_t *key, void **_schedule) {
src/crypto/aes/aes_c.c:void OQS_API oqs_aes256_free_schedule_c(void *schedule) {
src/crypto/aes/aes_c.c:void OQS_API oqs_aes128_enc_c(const uint8_t *plaintext, const void *_schedule, uint8_t *ciphertext) {
src/crypto/aes/aes_c.c:void OQS_API oqs_aes128_dec_c(const uint8_t *ciphertext, const void *_schedule, uint8_t *plaintext) {
src/crypto/aes/aes_c.c:void OQS_API oqs_aes256_enc_c(const uint8_t *plaintext, const void *_schedule, uint8_t *ciphertext) {
src/crypto/aes/aes_c.c:void OQS_API oqs_aes256_dec_c(const uint8_t *ciphertext, const void *_schedule, uint8_t *plaintext) {
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_load_schedule_ni(const uint8_t *key, void **schedule);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_free_schedule_ni(void *schedule);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_enc_ni(const uint8_t *plaintext, const void *schedule, uint8_t *ciphertext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_dec_ni(const uint8_t *ciphertext, const void *schedule, uint8_t *plaintext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_enc_ni(const uint8_t *plaintext, const size_t plaintext_len, const uint8_t *key, uint8_t *ciphertext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_dec_ni(const uint8_t *ciphertext, const size_t ciphertext_len, const uint8_t *key, uint8_t *plaintext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_enc_sch_ni(const uint8_t *plaintext, const size_t plaintext_len, const void *schedule, uint8_t *ciphertext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_dec_sch_ni(const uint8_t *ciphertext, const size_t ciphertext_len, const void *schedule, uint8_t *plaintext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_load_schedule_c(const uint8_t *key, void **schedule);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_free_schedule_c(void *schedule);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_enc_c(const uint8_t *plaintext, const void *schedule, uint8_t *ciphertext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_dec_c(const uint8_t *ciphertext, const void *schedule, uint8_t *plaintext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_enc_c(const uint8_t *plaintext, const size_t plaintext_len, const uint8_t *key, uint8_t *ciphertext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_dec_c(const uint8_t *ciphertext, const size_t ciphertext_len, const uint8_t *key, uint8_t *plaintext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_enc_sch_c(const uint8_t *plaintext, const size_t plaintext_len, const void *schedule, uint8_t *ciphertext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_dec_sch_c(const uint8_t *ciphertext, const size_t ciphertext_len, const void *schedule, uint8_t *plaintext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_load_schedule_ossl(const uint8_t *key, void **schedule, int for_encryption);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_free_schedule_ossl(void *schedule);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_enc_ossl(const uint8_t *plaintext, const size_t plaintext_len, const uint8_t *key, uint8_t *ciphertext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_dec_ossl(const uint8_t *ciphertext, const size_t ciphertext_len, const uint8_t *key, uint8_t *plaintext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_enc_sch_ossl(const uint8_t *plaintext, const size_t plaintext_len, const void *schedule, uint8_t *ciphertext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes128_ecb_dec_sch_ossl(const uint8_t *ciphertext, const size_t ciphertext_len, const void *schedule, uint8_t *plaintext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes256_load_schedule_c(const uint8_t *key, void **schedule);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes256_free_schedule_c(void *schedule);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes256_enc_c(const uint8_t *plaintext, const void *schedule, uint8_t *ciphertext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes256_dec_c(const uint8_t *ciphertext, const void *schedule, uint8_t *plaintext);
src/crypto/aes/aes_local.h:void OQS_API oqs_aes256_ecb_enc_c(const uint8_t *plaintext, const size_t plaintext_len, const uint8_t *key, uint8_t *ciphertext);

**Comment by smashra** [comment]
The list of functions I had to prefix with OQS_API for compilation to work
is attached to the PR# 363.

**Comment by dstebila** [comment]
Looking at this list, the only ones that I think we would want to consider part of our public API are:
```
OQS_MEM_cleanse
OQS_MEM_secure_free
OQS_MEM_insecure_free
OQS_randombytes_nist_kat_init
```

Some of the `OQS_RAND_` ones will be removed as part of #389.  All of the AES stuff should definitely not be part of the public API.  How can we manage this?

**Comment by dstebila** [comment]
The functions in my previous comment will be made part of the public API.

But the rest of the flagged functions should still be not visible.  The reason these errors are occurring is that some of the test programs use the internal functions directly; we will change the building of the test programs (e.g., test\_aes) so that those are only built statically.  But we still want the main test programs (e.g., test\_kem, kat\_kem, example\_kem, same for sig) to be built and tested against both static and shared libraries.

**Pull request merged**